### PR TITLE
Support endpoints through env

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Support endpoints through env

## Motivation and Context
Support endpoints through env

## How Has This Been Tested?
Manually

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by balamurugana** [comment]
```
Error running mint automation
+ mkdir -p 5952-560df4e/gopath/src/github.com/minio
+ git -C 5952-560df4e/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 5952-560df4e/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 5952-560df4e/gopath/src/github.com/minio/minio fetch --quiet minio pull/5952/head:pr5952
+ git -C 5952-560df4e/gopath/src/github.com/minio/minio checkout --quiet pr5952
+ GOPATH=/home/bala/mintauto/5952-560df4e/gopath
+ make -C 5952-560df4e/gopath/src/github.com/minio/minio --quiet
++ minikube docker-env
+ eval 'export DOCKER_TLS_VERIFY="1"
export DOCKER_HOST="tcp://:2376"
export DOCKER_CERT_PATH="/home/bala/.minikube/certs"
export DOCKER_API_VERSION="1.23"
# Run this command to configure your shell:
# eval $(minikube docker-env)'
++ export DOCKER_TLS_VERIFY=1
++ DOCKER_TLS_VERIFY=1
++ export DOCKER_HOST=tcp://:2376
++ DOCKER_HOST=tcp://:2376
++ export DOCKER_CERT_PATH=/home/bala/.minikube/certs
++ DOCKER_CERT_PATH=/home/bala/.minikube/certs
++ export DOCKER_API_VERSION=1.23
++ DOCKER_API_VERSION=1.23
+ cp -af Dockerfile.mint-auto 5952-560df4e/gopath/src/github.com/minio/minio/
+ docker build --quiet --tag minio/minio-5952-560df4e --file 5952-560df4e/gopath/src/github.com/minio/minio/Dockerfile.mint-auto 5952-560df4e/gopath/src/github.com/minio/minio
Cannot connect to the Docker daemon at tcp://localhost:2376. Is the docker daemon running?

```

**Comment by harshavardhana** [comment]
Tested with limits of how much memory can be allocated, I can see that there is no such limit.

```go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	var b strings.Builder
	for i := 100 * 1024 * 1024; i >= 1; i-- {
		fmt.Fprintf(&b, "%d", i)
	}
	os.Setenv("TEST", b.String())
}
```

File: cmd/server-main.go [hunk 1]
```diff
@@ -164,7 +179,7 @@ func init() {
 
 // serverMain handler called for 'minio server' command.
 func serverMain(ctx *cli.Context) {
-	if (!ctx.IsSet("sets") && !ctx.Args().Present()) || ctx.Args().First() == "help" {
+	if ctx.Args().First() == "help" && !endpointsPresent(ctx) {
```
Comment (kannappanr) [inline cmd/server-main.go]: Needs to be `if ctx.Args().First() == "help" || !endpointsPresent(ctx) {`

File: cmd/server-main.go [hunk 2]
```diff
@@ -112,6 +111,16 @@ EXAMPLES:
 `,
 }
 
+// Checks if endpoints are either available through environment
+// or command line, returns false if both fails.
+func endpointsPresent(ctx *cli.Context) bool {
+	_, ok := os.LookupEnv("MINIO_ENDPOINTS")
+	if !ok {
+		ok = ctx.Args().Present()
```
Comment (kannappanr) [inline cmd/server-main.go:119]: Should this be `ok = ctx.Args().Present() || ctx.IsSet("sets")`

Comment (harshavardhana) [reply cmd/server-main.go:119]: There is no such thing as sets - that was a mistake.

File: cmd/server-main.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/server-main.go]: oh yes.

Comment (harshavardhana) [reply cmd/server-main.go]: actually yes, if help is provided we should show help anyways.

**Review by kannappanr — APPROVED** [review]
tested. LGTM

**Comment by harshavardhana** [comment]
Moving this to blocked until final :+1:  by @abperiasamy

**Comment by harshavardhana** [comment]
> Moving this to blocked until final  by @abperiasamy

@abperiasamy approved this.

File: cmd/server-main.go [hunk 3]
```diff
@@ -65,6 +66,9 @@ ENVIRONMENT VARIABLES:
      MINIO_ACCESS_KEY: Custom username or access key of minimum 3 characters in length.
      MINIO_SECRET_KEY: Custom password or secret key of minimum 8 characters in length.
 
+  ENDPOINTS:
+     MINIO_ENDPOINTS: List of all the remote endpoints delimited by ' '.
```
Comment (nitisht) [inline cmd/server-main.go]: Since even local endpoints will work AFAICS, shouldn't this be `List of all the endpoints delimited by ' '.`

**Review by vadmeste — APPROVED** [review]
LGTM & tested

**Pull request merged**